File: leetcode/1496/solution.go
/**
 * https://leetcode.com/problems/path-crossing/
 * 1496. Path Crossing
 */
package main

func isPathCrossing(path string) bool {
	x, y := 0, 0
	seen := make(map[Tuple]struct{})
	seen[Tuple{x, y}] = struct{}{}
	for _, c := range path {
		switch c {
		case 'N':
			y--
		case 'S':
			y++
		case 'W':
			x--
		default:
			x++
		}
		if _, ok := seen[Tuple{x, y}]; ok {
			return true
		}
		seen[Tuple{x, y}] = struct{}{}
	}
	return false
}

type Tuple struct {
	a, b int
}

func (t Tuple) Hash() uint64 {
	return uint64(t.a) ^ uint64(t.b)
}
func (t Tuple) Equal(other Tuple) bool {
	return t.a == other.a && t.b == other.b
}

File: leetcode/1496/solution.rs
/**
 * https://leetcode.com/problems/path-crossing/
 * 1496. Path Crossing
 */
use std::collections::HashSet;

impl Solution {
    pub fn is_path_crossing(path: String) -> bool {
        let (mut x, mut y) = (0, 0);
        let mut seen = HashSet::new();
        seen.insert((x, y));
        for c in path.chars() {
            match c {
                'N' => y -= 1,
                'S' => y += 1,
                'E' => x += 1,
                _ => x -= 1,
            }
            if seen.contains(&(x, y)) {
                return true;
            }
            seen.insert((x, y));
        }

        false
    }
}
